include/continuable/detail/support/asio.hpp

```diff
@@ -0,0 +1,175 @@
+  }                                                                            \
+  }
+#endif
+
+#if defined(CTI_DETAIL_ASIO_HAS_NO_INTEGRATION)
+#error \
+  "First-class ASIO support for continuable requires the form of "\
+  "`async_result` with an `initiate` static member function, which was added " \
+  "in standalone ASIO 1.13.0 and Boost ASIO 1.70. Older versions can be " \
+  "integrated manually with `cti::promisify`."
+#endif
+
+#include <utility>
+
+namespace cti {
+namespace detail {
+namespace asio {
+
+#if defined(ASIO_STANDALONE)
+using error_code_t = ::asio::error_code;
+using error_cond_t = std::error_condition;
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+using exception_t = ::asio::system_error;
+#endif
+#else
+using error_code_t = ::boost::system::error_code;
+using error_cond_t = ::boost::system::error_condition;
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+using exception_t = ::boost::system::system_error;
+#endif
+#endif
+
+#if defined(CONTINUABLE_HAS_EXCEPTIONS)
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept {
+  promise.set_exception(std::make_exception_ptr(exception_t(std::move(e))));
+}
+#else
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept
+    ->decltype(promise.set_exception(std::move(e))) {
+  return promise.set_exception(std::move(e));
+}
+
+template <typename Promise>
+void promise_exception_helper(Promise& promise, error_code_t e) noexcept
+    ->decltype(promise.set_exception(error_cond_t(e.value(), e.category()))) {
+  promise.set_exception(error_cond_t(e.value(), e.category()));
+}
+
+#endif
+
+// Binds `promise` to the first argument of a continuable resolver, giving it
+// the signature of an ASIO handler.
+template <typename Promise>
+auto promise_resolver_handler(Promise&& promise) noexcept {
+  return [promise = std::forward<Promise>(promise)](
+             error_code_t e, auto&&... args) mutable noexcept {
+    if (e) {
+      promise_exception_helper(promise, std::move(e));
+    } else {
+      promise.set_value(std::forward<decltype(args)>(args)...);
+    }
+  };
+}
+
+// Helper struct wrapping a call to `cti::make_continuable` and, if needed,
+// providing an erased, explicit `return_type` for `async_result`.
+template <typename Signature>
+struct initiate_make_continuable;
+
+template <typename... Args>
+struct initiate_make_continuable<void(error_code_t, Args...)> {
+  using is_void_cti_t = std::integral_constant<bool, sizeof...(Args) == 0>;
+
+#if defined(CTI_DETAIL_ASIO_HAS_EXPLICIT_RET_TYPE_INTEGRATION)
+  using erased_return_type =
+      std::conditional_t<is_void_cti_t::value, cti::continuable<>,
+                         cti::continuable<Args...>>;
+#endif
+
+  template <typename Continuation, typename IsVoid = is_void_cti_t>
+  auto operator()(Continuation&& continuation,
+                  std::enable_if_t<IsVoid::value>* = 0) {
+    return cti::make_continuable<void>(
```

Comment:
Take a look at https://github.com/Naios/continuable/blob/d842c142685f4758fa3487126b5e68d4edf6fcf2/include/continuable/continuable-base.hpp#L851-L854 
The internal facilities allow to create a continuable where `<Args...>` are the actual arguments including `<>` for `void`:

With the following code you don't need a specialization for `void`:
```cpp
detail::base::attorney::create_from(
      std::forward<Continuation>(continuation),
      detail::util::identity<Args...>{},
      detail::util::ownership{});
```

Reply:
Ah thanks will definitely use this! As I wrote that facility I suspected the logic would have occurred to you somewhere, so it's nice to have you directing me to the relevant internals!

Reply:
I've implemented this change and also realized the `conditional_t` to provide `erased_return_type` was an unnecessary relic of my misunderstanding where I thought it was `cti::continuable<void>` rather than `cti::continuable<>`.